internal/olm/operator/tenancy.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package olm
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+
+	registryutil "github.com/operator-framework/operator-sdk/internal/util/operator-registry"
+
+	olmapiv1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1"
+	olmapiv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// Annotation containing a CSV's member OperatorGroup's name.
+	olmOperatorGroupAnnotation = "olm.operatorGroup"
+	// Annotation containing a CSV's member OperatorGroup's namespace.
+	olmOperatorGroupNamespaceAnnotation = "olm.operatorNamespace"
+)
+
+// operatorGroupDown handles logic to decide whether the SDK-managed
+// OperatorGroup can be created.
+//
+// Check if an OperatorGroup needs to be created in m.namespace first. If
+// there is and another OpreatorGroup is created, CSV installation will fail
+// with reason TooManyOperatorGroups. If the CSV's installModes don't support
+// the target namespace selection of the OperatorGroup, the CSV will fail
+// with UnsupportedOperatorGroup.
+//
+// https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/operatorgroups.md
+func (m *operatorManager) operatorGroupUp(ctx context.Context) error {
```

Comment:
@njhale @kevinrizza does this OperatorGroup logic make sense? I'm doing a thorough writeup of this logic and push that soon, which will further clarify my thinking.

Reply:
Proposed logic: #2324